Azure/k8s-infra
# Add devcontainer for easy setup

This is based off other devcontainers that I've used for go in [azbrowse](https://github.com/lawrencegripper/azbrowse/pull/201) - (thanks to @stuartleeks who helped put the origins of this together with me.) 

To test out:
1. Pull the branch 
2. Install [VSCode Remote - Containers](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers) extension
3. `CTRL-Shift-P` -> `Remote-Container: Reopen in container` 

Once the docker build is done (only happens first time round) you should have a golang VSCode configured to use the google language server, kubebuilder, kubectl, kustomize, userful extensions etc all setup. 

Where possible in the devcontainer I've pinned versions so different people using the container should have the same `kubebuilder` version etc. 

I've also added a set of useful `make` targets which let you quickly deploy the controller into a local `kind` cluster from the devcontainer. For example `make kind-deploy` should start a `kind` cluster, build the controller, load the built image into the cluster and then deploy the k8s yaml for the controller into the cluster. **These don't work at the moment as targets like `make docker-build` will fail as we haven't added any `kinds` to the kubebuilder project yet** (expected paths like `./api` are empty) so needs more testing once that's in place.

In terms of golang vscode tooling I've kept things light as most of the extended tools VSCode installs for golang aren't needed when using the language server. 

Bit late UK time so raising as a draft as I've rushed a bit and want to review in the morning to make sure I haven't made any silly mistakes.
